lib/pipeline.ml

```diff
@@ -0,0 +1,128 @@
+(* Check whether a variant is considered experimental.
+   If it is experimental we allow those builds to fail without
+   failing the overall build for a commit.
+*)
+let experimental_variant variant =
+  Astring.String.is_prefix ~affix:"macos-homebrew" variant
+
+let list_errors ~ok errs =
+  let groups =
+    (* Group by error message *)
+    List.sort compare errs
+    |> List.fold_left
+         (fun acc (msg, l) ->
+           match acc with
+           | (m2, ls) :: acc' when m2 = msg -> (m2, l :: ls) :: acc'
+           | _ -> (msg, [ l ]) :: acc)
+         []
+  in
+  Error
+    (`Msg
+      (match groups with
+      | [] -> "No builds at all!"
+      | [ (msg, _) ] when ok = 0 ->
+          msg (* Everything failed with the same error *)
+      | [ (msg, ls) ] ->
+          Fmt.str "%a failed: %s" Fmt.(list ~sep:(any ", ") string) ls msg
+      | _ ->
+          (* Multiple error messages; just list everything that failed. *)
+          let pp_label f (_, l) = Fmt.string f l in
+          Fmt.str "%a failed" Fmt.(list ~sep:(any ", ") pp_label) errs))
+
+let summarise results =
+  results
+  |> List.fold_left
+       (fun (ok, pending, err, skip) -> function
+         | _, (Ok `Checked, _) ->
+             (ok, pending, err, skip) (* Don't count lint checks *)
+         | _, (Ok `Built, _) -> (ok + 1, pending, err, skip)
+         | l, (Error (`Msg m), _)
+           when Astring.String.is_prefix ~affix:"[SKIP]" m ->
+             (ok, pending, err, (m, l) :: skip)
+         | l, (Error (`Msg _), _) when experimental_variant l ->
+             (ok + 1, pending, err, skip)
+         (* Don't fail the commit if an experimental build failed. *)
+         | l, (Error (`Msg m), _) -> (ok, pending, (m, l) :: err, skip)
+         | _, (Error (`Active _), _) -> (ok, pending + 1, err, skip))
+       (0, 0, [], [])
+  |> fun (ok, pending, err, skip) ->
+  if pending > 0 then Error (`Active `Running)
+  else
+    match (ok, err, skip) with
+    | 0, [], skip ->
+        list_errors ~ok:0
+          skip (* Everything was skipped - treat skips as errors *)
+    | _, [], _ -> Ok () (* No errors and at least one success *)
+    | ok, err, _ -> list_errors ~ok err (* Some errors found - report *)
+
+open Current.Syntax
+module Git = Current_git
```

Comment:
Shouldn't this come at the top of the file?

Reply:
Introducing it here limits the scope for those symbols.